engine/src/main/java/org/hibernate/validator/internal/metadata/provider/AnnotationMetaDataProvider.java

```diff
@@ -696,27 +698,66 @@ else if ( constraintHelper.isMultiValueConstraint( annotationType ) ) {
 		}
 		AnnotatedType annotatedType = executable.getAnnotatedReturnType();
 
-		List<TypeVariable<?>> cascadingTypeParameters = getCascadingTypeParameters( typeParameters, annotatedType );
+		List<CascadingTypeParameter> cascadingTypeParameters = getCascadingTypeParameters( typeParameters, annotatedType );
 
 		if ( executable.isAnnotationPresent( Valid.class ) ) {
-			cascadingTypeParameters.add( isArray ? ArrayElement.INSTANCE : AnnotatedObject.INSTANCE );
+			cascadingTypeParameters.add( isArray
+					? CascadingTypeParameter.arrayElement( ReflectionHelper.typeOf( executable ) )
+					: CascadingTypeParameter.annotatedObject( ReflectionHelper.typeOf( executable ) ) );
 		}
 
-		return cascadingTypeParameters;
+		return cascadingTypeParameters.isEmpty() ? Collections.emptyList() : cascadingTypeParameters;
 	}
 
-	private List<TypeVariable<?>> getCascadingTypeParameters(TypeVariable<?>[] typeParameters, AnnotatedType annotatedType) {
-		List<TypeVariable<?>> cascadingTypeParameters = new ArrayList<>();
+	private List<CascadingTypeParameter> getCascadingTypeParameters(TypeVariable<?>[] typeParameters, AnnotatedType annotatedType) {
```

Comment:
That method became quite long. The two cases it handles are good candidates for extracting two methods.